### ISSUE TITLE
Is there a way to order on relationship properties?

### ISSUE THREAD

**Issue by nrdobie** [neo4jrb/neo4j#697]
I'd like to be able to pull five of the latest items by a date property that is stored on a relationship. These items aren't guaranteed to be added to the database in chronological order so I can't just pull the last five items. Is there a function like `order` that can sort relationships or do I have to use a Cypher query?

Here are the models I have:

```ruby
class Console
  include Neo4j::ActiveNode

  property :name,        type: String
  property :manufacture, type: String

  validates_presence_of :name
  validates_presence_of :manufacture

  has_one :out, :successor,   model_class: Console
  has_one :in,  :predecessor, model_class: Console, type: 'SUCCESSOR'

  has_many :in, :games, rel_class: ReleasedFor, model_class: Game, unique: true

end

#####

class Game
  include Neo4j::ActiveNode

  property :title, type: String

  validates_presence_of :title

  has_many :out, :released_for, rel_class: ReleasedFor, model_class: Console, unique: true

end

#####

class ReleasedFor
  include Neo4j::ActiveRel

  from_class Game
  to_class   Console
  type       'RELEASED_FOR'

  property :date, type: Date
end
```

Here is the code I try running:

```ruby
console.games.order(date: :desc).limit(5).each { |g| puts g.title }
```

**Comment by subvertallchris** [comment]
There are no options the ActiveRel class or ActiveNode associations that'll do that for you. You could create a scope or a class method that'll do it, then you can call `console.ordered_games.each`.

```
def self.ordered_games
  all.games.order(date: :desc).limit(5)
end
```

Call `all` at the beginning hook into the scope. Even though it's defined as a class method, you'll be able to use it on instances or within a QueryProxy chain.

**Comment by nrdobie** [comment]
I was able to get it to work using `console.games.order("#{console.games.rel_var}.date DESC").limit(5).each { |g| puts g.title }`. Although it would be nice to have a function similar to `.rel_where` like `.rel_order` that would make it easier to do this.

**Comment by subvertallchris** [comment]
That's a really good idea and it'd be very easy to implement. I can
probably find a few minutes to do it tomorrow.

**Comment by cheerfulstoic** [comment]
I have to say I don't like this `rel_*` trend...  You don't have a `REL WHERE` in cypher, and I don't think we should have one in the ruby.

A few ideas:

```ruby
  .where({var_on_node: value1}, var_on_rel: value2)
  # same idea for .order, etc...

  .where(rel: {var_on_rel: value2}, node: {var_on_node: value1})

  .where(var_on_node: value1).rel.where(var_on_rel: value2)
  # Here maybe the `rel` method could take an argument which would determine the rel's value, so we wouldn't have two variables on the association method like we do now where sometimes you have to say `association(nil, :rel_var)`
```

**Comment by subvertallchris** [comment]
I'm not sure about that. By supporting scopes, associations, and class methods, QP already provides and allows methods that simplify common patterns. A `rel` method, usable in  `rel.where`, would cool, especially since it'd keep us from having to do explicit `rel_*` methods.

**Comment by nrdobie** [comment]
Would it be possible for the plugin to just "know" if the property was on the node or relationship? The obvious problem is how to handle conflicts when both the node and the relationship have a property of the same name. Maybe then having a syntax like `.where({var_on_node: value1}, {var_on_rel: value2})` alongside it so the user can then explicitly tell Neo4j which one to use when there is a conflict.

**Comment by cheerfulstoic** [comment]
Oh, right, that reminds me of my other suggestion.  If there's a nested hash argument, we assume that the key is specifying the cypher variable:

```ruby
  .where(node_var: {var_on_node: value1}, rel_var: {var_on_rel: value2})
```

The idea of just figuring out the variable is tempting, but you're right about conflicts.  I don't like having `where` mix node/rel properties unless you specify a separate argument, though.  I think they should either always be mixed, or always be separate.

I'm not sure I understand what you said @subvertallchris.  What aren't you sure about?

**Comment by subvertallchris** [comment]
Not sure about QP having to reflect Cypher. I don't think the user shouldn't have to define or know node/rel vars for simple operations.

**Comment by subvertallchris** [comment]
But I think that having a `rel` method to change the target of a filtering function is a good compromise.

**Comment by nrdobie** [comment]
My personal preference would be to use this syntax.

```ruby
.where({var_on_node: value1}, {var_on_rel: value2})
```

While you'll occasionally have times where you use `.where(nil, {var_on_rel: :desc})`. I think out of all the presented suggestions, it is the most straightforward and easy to remember.

My problem with the `rel` method is that it changes what the `where` and `order` methods are updating, so if you pass query chain around you may think you are setting a where on the node but actually be setting on the relationship.

**Comment by cheerfulstoic** [comment]
I agree about the chaining of the `rel` method and how you don't really have an indication of where you are.

I think the two argument solution (hah) is the best compromise.  Though to @subvertallchris's point, when it comes to defining `where`/`order` clauses on relationships, I think we're at the point where you want to know about what nodes and relationships are.

**Comment by nrdobie** [comment]
I can redo my pull request using the two argument method this weekend if we think that's the best direction to move in. I'll change `rel_where` to use the new style and can add a depreciation warning for people using it.

**Comment by subvertallchris** [comment]
Let's leave rel_where alone for now. I'm not sold on the two argument
solution.

**Comment by subvertallchris** [comment]
It's not that I don't think people should know what nodes and relationships are, it's that I don't think that they should have to set identifiers for them earlier in their query if they want to filter on something. `where(node_var: { prop: value }, rel_var: { prop: value })` requires this. That's fine for Core::Query, but why should it be necessary for QueryProxy?

As for two arguments, `rel_where` is easier to read and understand when quickly looking through code. If you want to add a second argument to `where` and adjust `rel_where` to rely on that, I'm fine with that. There must be a dozen methods in QueryProxy that don't match anything in Cypher but they form a DSL that's very pleasant, in my opinion.

I was thinking that `rel` would make the next method refer to the relationship identifier but it would go back to the next node var after that... or something. ;-) It wouldn't be a permanent context change, it would just set the target of the next QP method called to the relationship.

**Comment by cheerfulstoic** [comment]
Yeah, I'm fine with leaving it as it is for now
